geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/SlotAdvisor.java

```diff
@@ -0,0 +1,230 @@
+    initializeBucketsIfNecessary();
+
+    Map<String, List<Integer>> memberBuckets = new HashMap<>();
+    for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+      String memberId =
+          dataRegion.getRegionAdvisor().getBucketAdvisor(bucketId).getPrimary().getUniqueId();
+      memberBuckets.computeIfAbsent(memberId, k -> new ArrayList<>()).add(bucketId);
+    }
+
+    return memberBuckets;
+  }
+
+  public synchronized List<MemberBucketSlot> getBucketSlots() {
+    initializeBucketsIfNecessary();
+
+    for (int bucketId = 0; bucketId < REDIS_REGION_BUCKETS; bucketId++) {
+      updateBucketDetails(bucketId);
+    }
+
+    return Collections.unmodifiableList(memberBucketSlots);
+  }
+
+  private synchronized MemberBucketSlot updateBucketDetails(int bucketId) {
+    MemberBucketSlot mbs = null;
+    try {
+      Pair<String, Integer> hostPort = getHostPort(bucketId);
+
+      mbs = memberBucketSlots.get(bucketId);
+      if (mbs == null) {
+        mbs = new MemberBucketSlot(bucketId, hostPort.getLeft(), hostPort.getRight());
+        memberBucketSlots.set(bucketId, mbs);
+      } else {
+        mbs.setPrimaryIpAddress(hostPort.getLeft());
+        mbs.setPrimaryPort(hostPort.getRight());
+      }
+    } catch (Exception ex) {
+      logger.error("Unable to update bucket detail for bucketId: {}", bucketId, ex);
+    }
+
+    return mbs;
+  }
+
+  private void initializeBucketsIfNecessary() {
+    if (dataRegion.getDataStore() != null &&
+        dataRegion.getDataStore().getAllLocalBucketIds().isEmpty()) {
+      PartitionRegionHelper.assignBucketsToPartitions(dataRegion);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private Pair<String, Integer> getHostPort(int bucketId) {
+    String memberId =
+        dataRegion.getRegionAdvisor().getBucketAdvisor(bucketId).getPrimary().getUniqueId();
+
+    if (hostPorts.containsKey(memberId)) {
+      return hostPorts.get(memberId);
+    }
+
+    Set<DistributedMember> membersWithDataRegion = new HashSet<>();
+    for (PartitionMemberInfo memberInfo : getRegionMembers(dataRegion)) {
+      membersWithDataRegion.add(memberInfo.getDistributedMember());
+    }
+
+    ResultCollector<RedisMemberInfo, List<RedisMemberInfo>> resultCollector =
+        FunctionService.onMembers(membersWithDataRegion)
+            .execute(RedisMemberInfoRetrievalFunction.ID);
+
+    hostPorts.clear();
+
+    for (RedisMemberInfo memberInfo : resultCollector.getResult()) {
+      Pair<String, Integer> hostPort =
+          Pair.of(memberInfo.getHostAddress(), memberInfo.getRedisPort());
+      hostPorts.put(memberInfo.getMemberId(), hostPort);
+    }
+
+    if (!hostPorts.containsKey(memberId)) {
+      // There is a very tiny window where this might happen - a member was hosting a bucket and
+      // died before the fn call above could even complete.
+      throw new RuntimeException("Unable to retrieve host and redis port for member: " + memberId);
```

Comment:
This doesn't seem like a good user experience to get exceptions in this situation - should this trigger a retry?

Reply:
I've added a retry loop now. Probably even less likely to happen now that we're recreating the buckets.